dnn: LSTM optimisation

smbz opened the pull request:
This uses the AVX-optimised fastGEMM1T for matrix multiplications where available, instead of the standard cv::gemm.

fastGEMM1T is already used by the fully-connected layer.  This PR involves two minor modifications to allow use with the LSTM layer:
 - Use unaligned access.  I don't believe this involves any performance hit in on modern CPUs (Nehalem and Bulldozer onwards) in the case where the address is actually aligned.
 - Allow for weight matrices where the number of columns is not a multiple of 8.

I have not enabled AVX-512 or RVV as I don't have access to such CPUs to test on.  If there are no problems with this PR I'll send a separate PR which deals with ARM NEON.

<cut/>

Performance comparison (vs built-in gemm, i.e. build without a BLAS):

```
Median (ms)

                     Name of Test                      original4  AVX    AVX2     AVX        AVX2
                                                                                   vs         vs
                                                                               original4  original4
                                                                               (x-factor) (x-factor)
lstm::Layer_LSTM::BATCH=1, IN=64, HIDDEN=192, TS=100    22.851   3.819  3.093     5.98       7.39
lstm::Layer_LSTM::BATCH=1, IN=192, HIDDEN=192, TS=100   31.183   4.601  3.252     6.78       9.59
lstm::Layer_LSTM::BATCH=1, IN=192, HIDDEN=512, TS=100   144.946  19.302 15.570    7.51       9.31
lstm::Layer_LSTM::BATCH=1, IN=1024, HIDDEN=192, TS=100  93.162   10.639 8.068     8.76      11.55
lstm::Layer_LSTM::BATCH=64, IN=64, HIDDEN=192, TS=2     27.403   5.502  4.301     4.98       6.37
lstm::Layer_LSTM::BATCH=64, IN=192, HIDDEN=192, TS=2    40.038   6.830  5.548     5.86       7.22
lstm::Layer_LSTM::BATCH=64, IN=192, HIDDEN=512, TS=2    200.630  29.869 22.061    6.72       9.09
lstm::Layer_LSTM::BATCH=64, IN=1024, HIDDEN=192, TS=2   119.687  14.251 11.480    8.40      10.43
lstm::Layer_LSTM::BATCH=128, IN=64, HIDDEN=192, TS=2    57.895   10.394 9.101     5.57       6.36
lstm::Layer_LSTM::BATCH=128, IN=192, HIDDEN=192, TS=2   82.741   13.166 10.971    6.28       7.54
lstm::Layer_LSTM::BATCH=128, IN=192, HIDDEN=512, TS=2   394.310  60.755 45.158    6.49       8.73
lstm::Layer_LSTM::BATCH=128, IN=1024, HIDDEN=192, TS=2  245.459  28.773 23.413    8.53      10.48
```

### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or other license that is incompatible with OpenCV
- [x] The PR is proposed to proper branch
- [x] There is reference to original bug report and related work (N/A)
- [x] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name. (N/A)
- [x] The feature is well documented and sample code can be built with the project CMake


```
force_builders=Linux AVX2
```

alalek commented on modules/dnn/src/layers/layers_common.simd.hpp:
```diff
@@ -557,6 +557,16 @@ void fastGEMM1T( const float* vec, const float* weights,
 {
     int i = 0;
 
+    uint32_t tailMaskArray[15];
+    for (int i = 0; i < 8; i++) {
+        tailMaskArray[i] = 0;
+    }
+    for (int i = 8; i < 15; i++) {
+        tailMaskArray[i] = 0xffffffffUL;
+    }
```
use static preallocated/reinitialized buffer:`static uint32_t tailMaskArray[15] = { ... };`

alalek commented on modules/dnn/src/layers/layers_common.simd.hpp:
```diff
@@ -613,7 +651,6 @@ void fastGEMM1T( const float* vec, const float* weights,
     _mm256_zeroupper();
 }
 
-
```
please avoid unrelated whitespace changes

alalek commented on modules/dnn/src/layers/recurrent_layers.cpp:
```diff
@@ -359,9 +365,72 @@ class LSTMLayerImpl CV_FINAL : public LSTMLayer
+                    }
+                else
+#endif
                 gemm(hInternal, Wh, 1, gates, 1, gates, GEMM_2_T);  //+Wh * h_{t-1}
```
Please add braces and indentation for gemm call:
```
{
    gemm(...)
}
```

alalek commented on modules/dnn/src/layers/recurrent_layers.cpp:
```diff
@@ -118,10 +120,14 @@ class LSTMLayerImpl CV_FINAL : public LSTMLayer
     LSTMLayerImpl(const LayerParams& params)
-        : numTimeStamps(0), numSamples(0)
+        : numTimeStamps(0), numSamples(0),
+          useAVX(checkHardwareSupport(CPU_AVX)),
+          useAVX2(checkHardwareSupport(CPU_AVX2))
```
Please use conditional compilation:

```
#if CV_TRY_AVX
    , useAVX(checkHardwareSupport(CPU_AVX))
#endif
#if CV_TRY_AVX2
    , useAVX2(checkHardwareSupport(CPU_AVX2))
#endif
```

alalek reviewed (commented):
Thank you for contribution!

Please take a look on failed tests on [Linux AVX2](http://pullrequest.opencv.org/buildbot/builders/precommit_linux64-avx2/builds/1946) builder ("default" checks).

smbz commented:
@alalek Thanks for the feedback - I think I've addressed all the comments.

The failed test was pointing to an memory access issue caused by how the new tail handling interacted with the padding logic from the fully connected layer if the number of input channels was < 8.  Now fixed.

There was also a potential issue where Infs in the weight matrix could be incorrectly converted to NaNs because of the multiplication by zero in the tail.  Fixed by masking both the weights and the input vector.

It looks like the win64 build has failed before actually building the code - is there a way to re-run it without a new commit?

JulieBar commented on modules/dnn/src/layers/recurrent_layers.cpp:
```diff
@@ -359,9 +369,74 @@ class LSTMLayerImpl CV_FINAL : public LSTMLayer
-                gemm(hInternal, Wh, 1, gates, 1, gates, GEMM_2_T);  //+Wh * h_{t-1}
-                gemm(dummyOnes, bias, 1, gates, 1, gates);          //+b
+#if CV_TRY_AVX2
+                if (useAVX2 && xCurr.isContinuous() && gates.isContinuous() && bias.isContinuous() && Wx.depth() == CV_32F && xCurr.depth() == CV_32F && gates.depth() == CV_32F && bias.depth() == CV_32F && Wx.cols >= 8)
```
could you please improve readability of the code as it does at the [fully connected layer](https://github.com/opencv/opencv/blob/master/modules/dnn/src/layers/fully_connected_layer.cpp#L231)? Condition 
```c++
gates.isContinuous() && bias.isContinuous() && gates.depth() == CV_32F && bias.depth() == CV_32F
```
could be calculated once. Or you can add a helper to ensure that Mats are continuous and have 32F type.

JulieBar commented on modules/dnn/src/layers/recurrent_layers.cpp:
```diff
@@ -359,9 +369,74 @@ class LSTMLayerImpl CV_FINAL : public LSTMLayer
+#endif
+                {
+                    gemm(hInternal, Wh, 1, gates, 1, gates, GEMM_2_T);  //+Wh * h_{t-1}
+                }
```
is it possible to add optimization in case of peephole too?
https://github.com/opencv/opencv/blob/master/modules/dnn/src/layers/recurrent_layers.cpp#L377

asmorkalov commented:
@smbz Friendly reminder.

smbz commented:
Thanks for the reminder!  Just to say that this isn't dead, I should get round to it in the next few days.

asmorkalov commented:
@smbz Friendly reminder.

alalek reviewed (approved):
Well done! Thank you for the contribution :+1:
